bip-0360.mediawiki

```diff
@@ -0,0 +1,608 @@
+version of Taproot.
+
+=== Transaction Serialization ===
+
+Following BIP-141, a new transaction serialization format is introduced to include an attestation field after the witness field:
+
+  [nVersion][marker][flag][txins][txouts][witness][attestation][nLockTime]
+
+* <code>marker</code>: <code>0x00</code> (same as SegWit)
+
+* <code>flag</code>: <code>0x02</code> (indicates the presence of both witness and attestation data)
+
+* <code>attestation</code>: Contains the quantum-resistant public keys and signatures.
+
+=== Attestation Structure ===
+
+The attestation field consists of:
+
+* <code>num_pubkeys</code>: The number of public keys
+([https://learnmeabitcoin.com/technical/general/compact-size/ compact size]).
+
+For each public key:
+
+* <code>pubkey_length</code>: compact size length of the public key.
+* <code>pubkey</code>: The public key bytes.
+
+Then:
+
+* <code>num_signatures</code>: The number of signatures (compact size).
+
+For each signature:
+
+* <code>signature_length</code>: compact size length of the signature.
+* <code>signature</code>: The signature bytes.
+
+This structure repeats for each input, in order, for flexibility in supporting multisig schemes and various
+quantum-resistant algorithms.
+
+For each input, a separate attestation field is used. To know how many attestation fields are present, implementations
+must count the number of inputs present in the transaction.
+
+=== Signature Algorithm Identification ===
+
+The specific quantum-resistant signature algorithm used is inferred from the length of the public key.
+Implementations must recognize the supported algorithms and validate accordingly.
+
+Supported PQC algorithms and their NIST Level V parameters:
+
+* '''SPHINCS+-256f:'''
+  * Public Key Length: 64 bytes
+  * Signature Length: 49,856 bytes
+* '''CRYSTALS-Dilithium Level 5:'''
+  * Public Key Length: 2,592 bytes
+  * Signature Length: 4,595 bytes
+* '''FALCON-1024:'''
+  * Public Key Length: 1,793 bytes
+  * Signature Length: 1,280 bytes
+* '''SQIsign NIST-V:'''
+  * Public Key Length: 128 bytes
+  * Signature Length: 335 bytes
+
+Implementations must reject public keys and signatures that do not match expected lengths for supported algorithms.
+
+=== Script Validation ===
+
+To spend a P2QRH output, the following conditions must be met:
+
+1. The <code>scriptPubKey</code> must be of the form:
+
+OP_PUSHNUM_3 <32-byte hash>
+
+2. The attestation must include:
+
+* The quantum-resistant public key(s) whose HASH256 concatenated and hashed again matches the <nowiki><hash></nowiki> in
+the <code>scriptPubKey</code>.
+
+* Valid signatures corresponding to the public key(s) and the transaction data.
+
+3. For multi-signature schemes, all required public keys and signatures must be provided for that input within the
+attestation. Public keys that are not needed can be excluded by including their hash in the attestation accompanied
+with an empty signature. This includes classical Schnorr signatures.
```

Comment:
For a multisignature scheme, would you need to reveal multiple leafs from the pubkey tree? From what I understood the tree can only hold public keys, not scripts. How then is the threshold communicated? Wouldn’t a spender be able to reveal only their own key and provide a signature for that?

Reply:
That's a good point. Does that need to be committed to in the output or just expressed in the attestation?

Reply:
If it’s not committed to in advance, you are building either a 1-of-n or a 0-of-n scheme, depending on the minimum value for the threshold.

Reply:
That makes sense. So, we essentially need to commit to a script hash. I'm thinking we just do a bunch of consecutive data pushes of PKHs and they correspond to leaves on a binary tree. This is then included in the witness. Keys in the attestation are hashed and compared to the PKHs in the v3 witness. Like this:

```
OP_3
OP_PUSHBYTES_32
04d81fd577272bbe73308c93009eec5dc9fc319fc1ee2e7066e17220a5d47a1
OP_PUSHBYTES_32
8314578be2faea34b9f1f8ca078f8621acd4bc22897b03daa422b9bf56646b3
OP_PUSHBYTES_32
04ec3afff0b2b66e8152e9018fe3be3fc92b30bf886b3487a525997d00fd9da
OP_PUSHBYTES_32
2d012dce5d5275854adc3106572a5d1e12d4211b228429f5a7b2f7ba92eb047
OP_PUSHBYTES_32
04b49b496684b02855bc32f5daefa2e2e406db4418f3b86bca5195600951c7d
OP_5
OP_CHECKMULTISIG
```

Notice, all 5 PKHs need to be committed to in advance in the script hash. Maybe we need to introduce a concept like QPKH? QuBit public key hash? And QSH for script hashes? Or would APKH / ASH be better, for attestation?

What do you think? This I think will obviate the necessity for a merkle tree, as recommended by @EthanHeilman. If 3 public keys aren't included and don't hash to any of the public keys in the script hash, then the transaction fails.